### PR TITLE
OM changes based on proposed .4 version of SARIF

### PR DESCRIPTION
**Pull request by michaelcfanning**
@lgolding @BillyONeal @davidknise @bobschumaker @tikiwanMS @nguerrera 

Start updating SDK to current SARIF spec. We need to make an additional pass over this code (including getting the unit tests working against).

### CODE REVIEW

File: src/StaticAnalysisResultsInterchangeFormat/Converters/AndroidStudioConverter.cs [hunk 1]
```diff
@@ -15,7 +15,7 @@
 namespace Microsoft.CodeAnalysis.StaticAnalysisResultsInterchangeFormat.Converters
 {
     /// <summary>
-    /// Converts an xml log file of the Android Studio format into the SARIF OES issue format
+    /// Converts an xml log file of the Android Studio format into the SARIF OES result format
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/Converters/AndroidStudioConverter.cs]: I assume OES stood for "One Engineering System". Remove.

File: src/StaticAnalysisResultsInterchangeFormat/Converters/CppCheckError.cs [hunk 2]
```diff
@@ -109,16 +109,16 @@ public static CppCheckError Parse(XmlReader reader, CppCheckStrings strings)
...
-        /// <returns>This instance as an <see cref="Issue"/>.</returns>
-        public Issue ToSarifIssue()
+        /// <summary>Converts this instance to <see cref="Result"/>.</summary>
+        /// <returns>This instance as an <see cref="Result"/>.</returns>
+        public Result ToSarifIssue()
         {
             if (this.Locations.Length == 0)
             {
-                throw new InvalidOperationException("At least one location must be present in a Unified Issue Store issue.");
+                throw new InvalidOperationException("At least one location must be present in a Unified Result Store result.");
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/Converters/CppCheckError.cs]: "Unified Issue Store" was an MS-centric name. Change "in a Unified Result Store Result" to just "in a Result". Look for other occurrences of "OES", "1ES", "Unified Issue Store", and "UIS".

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 3]
```diff
@@ -1,15 +1,15 @@
-/** @namespace {Microsoft.CodeAnalysis.StaticAnalysisResultsInterchangeFormat.DataContracts} */ grammar ToolIssuesLog;
+/** @namespace {Microsoft.CodeAnalysis.StaticAnalysisResultsInterchangeFormat.DataContracts} */ grammar ToolResultsLog;
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: "ToolResultLog" isn't a thing.

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 4]
```diff
@@ -36,44 +36,44 @@ runLog:
...
-        The name of the tool or tool pipeline that generated the issues in this log, e.g., FxCop.
+        The name of the tool or tool pipeline that generated the results in this log, e.g., FxCop.
         }
     */
-    toolName
+    name
 
     /**
-        @name {FullVersion}
+        @name {FullName}
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: Comment is wrong. Spec says that `toolInfo.fullName` is "string containing the name of the tool along with its version and any other useful identifying information, such as its locale. "

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 5]
```diff
@@ -101,20 +104,20 @@ toolInfo :
...
 	NOTE: The information in the runInfo object makes it possible to precisely repeat a run of a
-	static analysis tool, and to verify that the issues reported in the log file were generated 
+	static analysis tool, and to verify that the results reported in the log file were generated 
 	by an appropriate invocation of the tool.
 	}
 */
 runInfo :
     /**
-		@name {CommandLineArguments}
+		@name {Parameterization}
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: `InvocationInfo`

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 6]
```diff
@@ -124,8 +127,8 @@ runInfo :
         @name {AnalysisTargets}
         @summary {
 		An array, each of whose elements is a fileReference object (�6.7) representing the location of 
-		a single analysis target scanned during the run. This array shall contain one entry for each 
-		analysis target that was scanned, even if the analysis targets were not individually specified 
+		a single analysis target scanned during the run. When presenct, this array shall contain one entry 
+		fo reach analysis target that was scanned, even if the analysis targets were not individually specified
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: "prescent" => "present"

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 7]
```diff
@@ -124,8 +127,8 @@ runInfo :
         @name {AnalysisTargets}
         @summary {
 		An array, each of whose elements is a fileReference object (�6.7) representing the location of
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: Should we include these paragraph references? They're not clickable here; how valuable are they?

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 8]
```diff
@@ -202,58 +205,69 @@ hash :
...
 
 /**
-    @className {Issue}
+    @className {Result}
     @summary {
     Represents one or more observations about an analysis target produced by a static analysis tool.
     }
     @remarks {
-    An issue frequently, but not always, represents a code defect.
+    An result frequently, but not always, represents a code defect.
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: "An result" => "A result"

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 9]
```diff
@@ -202,58 +205,69 @@ hash :
...
+        RuleIds need only be unique in the context of a specific tool, that is, a rule CA2001
+        defined by ToolA is distinct from a rule with the same ID defined by ToolB.
         }
     */
-    toolFingerprint?
+    ruleId
 
     /**
-        @name {RuleId}
+        @name {ResultKind}
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: We're using just plain "Kind" in each context, aren't we?

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 10]
```diff
@@ -356,10 +398,18 @@ location :
     */
     physicalLocation?
 
+	/**
+        @name {LogicalLocation} @serializedName {logicalLocation}
+        @summary {
+        An object that specifies the logical namespace / function / etc location.
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: I don't think this sentence parses right.

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 11]
```diff
@@ -619,11 +573,243 @@ region :
...
+    /**
+        @name {LocationKind}
+        @summary {
+        The type of item this location refers to.
+        }
+        @remarks {
+        Examples include namespace, class, or function.
+        }
+    */
+    locationKind;
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: Again, maybe just "kind" (and adjust spec accordingly")

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 [hunk 12]
```diff
@@ -619,11 +573,243 @@ region :
...
+/** @className {AnnotatedCodeLocations}
+    @summary {
+    A set of one or more annotated code locations associated with a given result.
+    }
+*/
+annotatedCodeLocations: annotatedCodeLocation*;
+
+/*********************************************************
+Execution flow infrastructure; execution flows are the same lists of physical locations except that
+each physical location in the list can have a message stapled on.
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: I'm not fond of the tone of "stapled on", and in any case I'm not sure how valuable this whole sentence (which I know is pre-existing) is.

File: src/StaticAnalysisResultsInterchangeFormat/IResultLogWriter.cs [hunk 13]
```diff
@@ -0,0 +1,34 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.CodeAnalysis.StaticAnalysisResultsInterchangeFormat.DataContracts;
+
+namespace Microsoft.CodeAnalysis.StaticAnalysisResultsInterchangeFormat
+{
+    /// <summary>This interface serves as a sink for <see cref="IssueLog"/> format issues.</summary>
```
Comment (lgolding) [inline src/StaticAnalysisResultsInterchangeFormat/IResultLogWriter.cs]: `IssueLog` => `ResultLLog`, except I think the whole sentence should be "Interface that serves as a sink for SARIF results"

**Comment by lgolding** [comment]
In a separate PR, I'd like to see the namespace changed to `Microsoft.CodeAnalysis.Sarif`. The full name is just too verbose for code, IMO.

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 (hunk 11 above)
Comment (michaelcfanning) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: Actually, there is a problem here: the grammar processor puts a default item named Kind on each class, so no other type can define a member with this name. :) I'll open an issue on this.

File: src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4 (hunk 3 above)
Comment (michaelcfanning) [inline src/StaticAnalysisResultsInterchangeFormat/GrammarFiles/SarifFormat.g4]: right, this is just an identifier for the grammar. Changed it to SarifGrammar.

**Comment by michaelcfanning** [comment]
@lgolding thanks for the review. fixed things up per your feedback. will address namespace change in a subsequent PR.

**Pull request merged**